File: agent/linux/pkg/evidence/EvidenceRecorder.go
package evidence

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"
)

const evidenceRoot = "/var/lib/fenrir/evidence"

type EvidenceBundle struct {
	BundleId    string    `json:"bundleId"`
	Pid         uint32    `json:"pid"`
	ProcessName string    `json:"processName"`
	CreatedAt   time.Time `json:"createdAt"`
	ProcessTree string    `json:"processTree,omitempty"`
	OpenFiles   string    `json:"openFiles,omitempty"`
	NetConns    string    `json:"netConnections,omitempty"`
}

type EvidenceRecorder struct{}

func NewEvidenceRecorder() *EvidenceRecorder {
	os.MkdirAll(evidenceRoot, 0700)
	return &EvidenceRecorder{}
}

// Collect gathers forensic evidence for a suspicious process.
func (e *EvidenceRecorder) Collect(pid uint32, processName string) (*EvidenceBundle, error) {
	bundleId := fmt.Sprintf("%d_%d", time.Now().UnixNano(), pid)
	bundle := &EvidenceBundle{
		BundleId:    bundleId,
		Pid:         pid,
		ProcessName: processName,
		CreatedAt:   time.Now(),
	}

	pidStr := fmt.Sprintf("%d", pid)

	if out, err := exec.Command("pstree", "-p", pidStr).Output(); err == nil {
		bundle.ProcessTree = string(out)
	}
	if out, err := exec.Command("lsof", "-p", pidStr).Output(); err == nil {
		bundle.OpenFiles = string(out)
	}
	if out, err := exec.Command("ss", "-tp", fmt.Sprintf("( dport %s or sport %s )", pidStr, pidStr)).Output(); err == nil {
		bundle.NetConns = string(out)
	}

	bundlePath := filepath.Join(evidenceRoot, bundleId+".json")
	data, _ := json.MarshalIndent(bundle, "", "  ")
	if err := os.WriteFile(bundlePath, data, 0600); err != nil {
		return bundle, fmt.Errorf("write evidence: %w", err)
	}

	return bundle, nil
}

File: agent/macos/Sources/FenrirAgent/EvidenceRecorder.swift
import Foundation

struct EvidenceBundle: Codable {
    let bundleId: String
    let pid: pid_t
    let processName: String
    let createdAt: String
    let processTree: String?
    let openFiles: String?
    let netConnections: String?
}

class EvidenceRecorder {
    private let evidenceRoot: URL

    init() {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .localDomainMask).first!
        evidenceRoot = base.appendingPathComponent("Fenrir/Evidence", isDirectory: true)
        try? FileManager.default.createDirectory(at: evidenceRoot, withIntermediateDirectories: true)
    }

    func collect(pid: pid_t, processName: String) -> EvidenceBundle {
        let bundleId = "\(Date().timeIntervalSince1970)_\(pid)"
        let pidStr = "\(pid)"

        let bundle = EvidenceBundle(
            bundleId: bundleId,
            pid: pid,
            processName: processName,
            createdAt: ISO8601DateFormatter().string(from: Date()),
            processTree: run("pstree", args: [pidStr]),
            openFiles: run("lsof", args: ["-p", pidStr]),
            netConnections: run("lsof", args: ["-i", "-nP", "-p", pidStr])
        )

        let bundlePath = evidenceRoot.appendingPathComponent("\(bundleId).json")
        if let data = try? JSONEncoder().encode(bundle) {
            try? data.write(to: bundlePath, options: .atomic)
        }

        print("[EvidenceRecorder] Saved evidence bundle: \(bundleId)")
        return bundle
    }

    private func run(_ command: String, args: [String]) -> String? {
        let process = Process()
        // Try common system binary paths
        for prefix in ["/usr/bin", "/bin", "/usr/sbin"] {
            let path = "\(prefix)/\(command)"
            if FileManager.default.isExecutableFile(atPath: path) {
                process.executableURL = URL(fileURLWithPath: path)
                break
            }
        }
        guard process.executableURL != nil else { return nil }
        process.arguments = args
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = Pipe() // suppress errors
        try? process.run()
        process.waitUntilExit()
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        return String(data: data, encoding: .utf8)
    }
}
